src/mam4xx/mam4_amicphys.hpp

```diff
@@ -0,0 +1,2371 @@
+  // qnum_cur(max_mode)
+  // qaer_cur(1:max_aer,1:max_mode)
+  // qwtr_cur(1:max_mode)
+
+  // DESCRIPTION:
+  //   computes changes due to aerosol nucleation (new particle formation)
+  //       treats both nucleation and subsequent growth of new particles
+  //          to aitken mode size
+  //   uses the following parameterizations
+  //       vehkamaki et al. (2002) parameterization for binary
+  //           homogeneous nucleation (h2so4-h2o) plus
+  //       kerminen and kulmala (2002) parameterization for
+  //           new particle loss during growth to aitken size
+  //
+  // REVISION HISTORY:
+  //   R.Easter 2007.09.14:  Adapted from MIRAGE2 code and CMAQ V4.6 code
+  //
+
+  constexpr int newnuc_method_flagaa = 11;
+  //  1=merikanto et al (2007) ternary   2=vehkamaki et al (2002) binary
+  // 11=merikanto ternary + first-order boundary layer
+  // 12=merikanto ternary + second-order boundary layer
+
+  // begin
+  dnclusterdt = 0;
+
+  // qh2so4_cur = current qh2so4, after aeruptk
+  // qh2so4_avg = average qh2so4 over time-step
+  // BAD CONSTANTS
+  constexpr Real qh2so4_cutoff = 4.0e-16;
+  Real qh2so4_cur = qgas_cur[igas_h2so4];
+  Real qh2so4_avg, tmp_uptkrate, tmpa;
+
+  // use qh2so4_avg and first-order loss rate calculated in
+  // mam_gasaerexch_1subarea
+  qh2so4_avg = qgas_avg[igas_h2so4];
+  tmp_uptkrate = uptkrate_h2so4;
+
+  if (qh2so4_avg <= qh2so4_cutoff)
+    return;
+
+  static constexpr int igas_nh3 = -999888777; // Same as mam_refactor
+  Real qnh3_cur = 0;
+
+  //   dry-diameter limits for "grown" new particles
+  constexpr int nait = static_cast<int>(ModeIndex::Aitken);
+  Real dplom_mode = haero::exp(0.67 * haero::log(modes(nait).min_diameter) +
+                               0.33 * haero::log(modes(nait).nom_diameter));
+  Real dphim_mode = modes(nait).max_diameter;
+
+  //   mass1p_... = mass (kg) of so4 & nh4 in a single particle of diameter ...
+  //                (assuming same dry density for so4 & nh4)
+  //      mass1p_aitlo - dp = dplom_mode
+  //      mass1p_aithi - dp = dphim_mode
+  constexpr Real dens_so4a_host = 1770;
+  tmpa = dens_so4a_host * haero::Constants::pi / 6.0;
+  Real mass1p_aitlo = tmpa * (haero::pow(dplom_mode, 3.0));
+  Real mass1p_aithi = tmpa * (haero::pow(dphim_mode, 3.0));
+
+  //   limit RH to between 0.1% and 99%
+  Real relhumnn = haero::max(0.01, haero::min(0.99, relhum));
+
+  // BAD CONSTANTS (These should come from chemistry mechanism
+  // but it is fixed here for stay BFB)
+  constexpr Real mw_so4a_host = 115;
+  constexpr Real mwnh4 = 18;
+  constexpr Real mwso4 = 96;
+  // BAD CONSTANTS (These should come from Haero constants
+  // but it is fixed here for stay BFB)
+  constexpr Real rgas = 8.31446759100000;
+  constexpr Real avogadro = 6.022140000000000E+023;
+
+  int itmp;
+  Real qnuma_del, qso4a_del, qnh4a_del, qh2so4_del, qnh3_del, dens_nh4so4a;
+
+  //   call ... routine to get nucleation rates
+  // FIXME: I GOT ALL ZEROS...THIS IS NOT VALIDATED YET!!!!
```

Comment:
Unless something's changed, this does have a passing validation test. Maybe old comment?

Reply:
We can remove this comment.